### DISCUSSION TITLE
How to format a duration into a string?

### DISCUSSION THREAD

**Discussion by adzubla** [nushell/nushell#8317] [Q&A]
I hava a variable of duration type and need to convert it into the string format "00:15:51.534", but could not find any solution...

```
G:\proj\nu-script〉$d | describe                                                                    03/04/2023 12:44:05 duration
G:\proj\nu-script〉$d | to text                                                                     03/04/2023 12:44:09 15min 51sec 534ms
G:\proj\nu-script〉$d | into string                                                                 03/04/2023 12:44:15
Error: nu::shell::cant_convert (link)

  × Can't convert to string.
   ╭─[entry #166:1:1]
 1 │ $d | into string
   ·      ─────┬─────
   ·           ╰── can't convert duration to string
   ╰────
```

**Comment by fdncred** [comment]
You can try something like this
```
"00:15:51.534" | date format '%H:%M:%S'
00:15:51
```

**Comment by adzubla** [comment]
Thanks for the quick answer, but it is not what I'm asking for.

It should be more clear with this example:
```
> let d = ("951.534868sec" | into duration) 
> $d
15min 51sec 534ms
```

How can I convert `$d` to the string "00:15:51.534"?

**Reply by fdncred** [reply]
This is a painful way to do it, and it's not a good way, but for the number you have above, you could do 
```
"951.534868sec" | into duration | to text | parse '{min}min {sec}sec {ms}ms' | $"00:($in.min.0):($in.sec.0).($in.ms.0)"
00:15:51.534
```
We already know those components. One would think we'd have a parameter to spit that out but I don't think we do.

**Reply by adzubla** [reply]
Why its possible to apply `to text` on a duration type, but not `into string`?

```
G:\〉"951.534868sec" | into duration | to text                                                      03/06/2023 11:39:59 15min 51sec 534ms
G:\〉"951.534868sec" | into duration | into string                                                  03/06/2023 11:40:06 Error: nu::shell::cant_convert (link)

  × Can't convert to string.
   ╭─[entry #6:1:1]
 1 │ "951.534868sec" | into duration | into string
   ·                                   ─────┬─────
   ·                                        ╰── can't convert duration to string
   ╰────
```

My understanting is that every type should be convertible to string.

**Comment by rgwood** [comment]
I don't have a perfect solution for you, but you can convert a duration to nanoseconds with `into int`.

**Reply by adzubla** [reply]
It's funny that duration has a lot of useful operations, but it can't be used in many scenarios that I can think of. After manipulating a duration one needs to output it, and the default format it not appropriate in most situations.

Have you considered adding more operations to it, like formating and extracting  individual fields?

Converting an object to a string just to parse it again, does not looks like the "nu way"! (I would expect to do this in bash...)

**Comment by isti115** [comment]
I ended up with a duration after subtracting one datetime from another, and managed to get it back into a usable output format by the following (still unnecessarily convoluted, but less so then the previously suggested) pipeline:
```nu
❯ $ts
7min 43sec 535ms

❯ $ts | describe
duration

❯ $ts | into int | into datetime | format date "%H:%M:%S.%f"
00:07:43.535000000
```
